go/multipooler/manager/consensus_state.go

```diff
@@ -0,0 +1,242 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manager
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	clustermetadatapb "github.com/multigres/multigres/go/pb/clustermetadata"
+	multipoolermanagerdatapb "github.com/multigres/multigres/go/pb/multipoolermanagerdata"
+)
+
+// ConsensusState manages the in-memory and on-disk consensus state for this node.
+// It provides thread-safe access to voting state and caches the state in memory
+// to avoid repeated disk reads.
+type ConsensusState struct {
+	poolerDir string
+	serviceID *clustermetadatapb.ID
+
+	mu    sync.RWMutex
+	term  *multipoolermanagerdatapb.ConsensusTerm
+	dirty bool // true if in-memory state differs from disk
+}
+
+// NewConsensusState creates a new ConsensusState manager.
+// It does not load state from disk - call Load() to initialize.
+func NewConsensusState(poolerDir string, serviceID *clustermetadatapb.ID) *ConsensusState {
+	return &ConsensusState{
+		poolerDir: poolerDir,
+		serviceID: serviceID,
+		term:      nil,
+		dirty:     false,
+	}
+}
+
+// Load loads consensus state from disk into memory.
+// If the file doesn't exist, initializes with default values (term 0, no vote).
+// This method is idempotent - subsequent calls will reload from disk.
+func (cs *ConsensusState) Load() error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	term, err := GetTerm(cs.poolerDir)
+	if err != nil {
+		return fmt.Errorf("failed to load consensus term: %w", err)
+	}
+
+	cs.term = term
+	cs.dirty = false
+	return nil
+}
+
+// Save persists the current in-memory state to disk using atomic write.
+// Only writes if the state has been modified (dirty flag is set).
+func (cs *ConsensusState) Save() error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if !cs.dirty {
+		// No changes to persist
+		return nil
+	}
+
+	if cs.term == nil {
+		return fmt.Errorf("cannot save nil consensus term")
+	}
+
+	if err := SetTerm(cs.poolerDir, cs.term); err != nil {
+		return fmt.Errorf("failed to save consensus term: %w", err)
+	}
+
+	cs.dirty = false
+	return nil
+}
+
+// GetCurrentTerm returns the current term.
+// Returns 0 if state has not been loaded.
+func (cs *ConsensusState) GetCurrentTerm() int64 {
+	cs.mu.RLock()
+	defer cs.mu.RUnlock()
+
+	if cs.term == nil {
+		return 0
+	}
+	return cs.term.GetCurrentTerm()
+}
+
+// GetAcceptedLeader returns the candidate ID this node voted for in the current term.
+// Returns empty string if no vote has been cast.
+func (cs *ConsensusState) GetAcceptedLeader() string {
+	cs.mu.RLock()
+	defer cs.mu.RUnlock()
+
+	if cs.term == nil || cs.term.AcceptedLeader == nil {
+		return ""
+	}
+	return cs.term.AcceptedLeader.GetName()
+}
+
+// GetTerm returns a copy of the current consensus term.
+// Returns nil if state has not been loaded.
+func (cs *ConsensusState) GetTerm() *multipoolermanagerdatapb.ConsensusTerm {
```

Comment:
Reflecting on potential risks of updating the term in memory while exposing/exporting that through GetTerm() or GetCurrentTerm() when it's possible Save() failed and didn't persist the data to disk. I'm nervous about people reading, sharing, or acting on term data that didn't make it to disk.

Would it be any safer for ConsensusState to track both a `proposedTerm` and `durableTerm`, where the proposed term would be promoted to into the durable term only when `Save()` succeeds?

Reply:
That might get complicated. But this reminded me that we should retry saving to disk if it fails, so I am adding that.

Reply:
Ok, I think what we should do is only update the in-memory value after we are able to save to disk. Any failure to save should percolate up the stack and return from the RPC as an error. Callers will then need to retry.